Normalize saliency maps

schencej opened the pull request:
The saliency maps are now normalized inside the generator function. The maps are then resized and scaled by 255 so that a value of 1 will give a blue/red value of 255 in the visual saliency representation. This operation replaces the post-processing function.

brianhhu commented on examples/atari_deepRL_saliency.ipynb:
```diff
@@ -206,7 +206,7 @@
      "name": "stdout",
```
Since we are editing this notebook, can we remove the "please" in the header section when talking about Colab? I think this was one of the changes that Mary Elise had wanted to make.

"To run this notebook in Colab, please use the link below:" -> "To run this notebook in Colab, use the link below:"

brianhhu reviewed (commented):
Left a few minor comments, mainly about reconsidering the use of post-processing function. @Purg, can you also take a look? This PR does not add a new interface/implementation to the base package, but only updates the notebook.

Purg commented on examples/atari_deepRL_saliency.ipynb:
```diff
@@ -457,21 +457,6 @@
     "print_lock = threading.Lock() # Lock to control thread printing\n",
     "\n",
-    "\n",
-    "def postprocess_sal(sal, ff):\n",
-    "    \"\"\"\n",
-    "    Helper function for post-processing saliency maps, which includes resizing\n",
-    "    the saliency maps to (160,160) and multiplying them by a \"fudge factor\" ff\n",
-    "    as in the original implementation.\n",
-    "    \"\"\"\n",
-    "    pmax = sal.max()\n",
-    "    sal = cv2.resize(src=sal, dsize=(160,160)).astype(np.float32)\n",
-    "    sal -= sal.min()\n",
-    "    sal = ff * pmax * sal / sal.max()\n",
-    "    \n",
-    "    return sal\n",
```
I thought we were trying to get rid of fudge-factors and the like with this revision? If you think it best to keep those in here for the sake of visualization, then OK, but we should be clear that this function is for that only. Ideally we would have a good explanation, or a link to one, for the effect of the fudge factor on the visualization, otherwise I don't know if it's just creating an illusions of success.

brianhhu replied:
I don't think it's possible to remove the fudge factor and still be faithful to the original implementation. However, I do believe that what Elim has now looks OK (without adding a fudge factor), so I'm happy to go with that if we don't think it's critical to match the original implementation exactly.

brianhhu commented on examples/atari_deepRL_saliency.ipynb:
```diff
@@ -457,21 +457,6 @@
-    "    return sal\n",
-    "\n",
-    "\n",
     "class PolicyScoring(GenerateClassifierConfidenceSaliency):\n",
     "    \"\"\"\n",
     "    This saliency implementation transforms black-box action and value scores\n",
```
We need a new implementation here since the saliency algorithm (proposed by Greydanus et al., '18) takes a squared difference between reference and perturbed scores, which is something that doesn't currently exist in our other classifier implementations.

schencej replied:
To quote the authors:

> We are interested in answering the question, “How much
does removing information from the region around location
(i, j) change the policy?"

They wanted to visualize the overall effect without discerning between positive and negative. They chose squared different for this.

schencej commented:
I've added the `PolicyScoring` implementation into the toolkit and changed the example to use that instead. As we discussed briefly in the meeting, I think an implementation that uses squared difference could apply to more than just policy networks. Perhaps we should rename and re-document.

I also changed the visualization a bit in the example. It's now handled by a separate function to emphasize that everything that is done to the saliency maps is purely for visualization. The maps are now alpha blended with the game frames instead of manipulating each frame's pixel intensities directly. This results in similar visualizations and doesn't require as much hacky stuff.

brianhhu commented on docs/implementations.rst:
```diff
@@ -48,6 +48,11 @@ Class: RISEScoring
 .. autoclass:: xaitk_saliency.impls.gen_classifier_conf_sal.rise_scoring.RISEScoring
   :members:
 
+Class: PolicyScoring
```
I think we decided to list implementations in alphabetical order, so please adjust this order appropriately based on whatever final name for the implementation we decide upon.

brianhhu commented on xaitk_saliency/impls/gen_classifier_conf_sal/policy_scoring.py:
```diff
@@ -0,0 +1,53 @@
+
+        return sal
+
+    def get_config(self) -> dict:
```
Looks like codecov complains about a missing test for the configuration portion of this new implementation- I think this should be relatively straightforward to add using a similar test from another implementation.

brianhhu commented on examples/atari_deepRL_saliency.ipynb:
```diff
@@ -83,6 +83,14 @@
+     "name": "stderr",
+     "output_type": "stream",
+     "text": [
+      "A.L.E: Arcade Learning Environment (version +a54a328)\n",
+      "[Powered by Stella]\n"
```
I noticed that this new output was added- are you still using the pinned version of `open-gym` (v. 0.19.0)? I ran into some issues with installing newer versions of `open-gym`, as I believe they updated how the ATARI games were provided with the installation in more recent versions. As a result, it looks like the generated frames/videos are also different in this update compared to those in the original notebook. I purposefully chose frames/clips (taking advantage of the determinism of setting a seed) that highlighted the "tunneling" strategy of the agent.

schencej replied:
Yeah, I was using 0.20.0. Fixed.

brianhhu commented on examples/atari_deepRL_saliency.ipynb:
```diff
@@ -600,15 +582,9 @@
-    "        # Highlight actor saliency in blue and critic saliency in red\n",
-    "        img[35:195,:,2] += postprocess_sal(actor_sal_maps[sal_idx], 300).astype('uint16')\n",
-    "        img[35:195,:,0] += postprocess_sal(critic_sal_maps[sal_idx], 600).astype('uint16')\n",
+    "        frame = visualize_saliency(history['ins'][img_idx], actor_sal_maps[sal_idx], critic_sal_maps[sal_idx])\n",
```
Just to confirm, do we need to create a copy of this frame history, or is it OK to pass in the original input here? We seem to create a copy when visualizing the individual frame, but not when generating the video.

e.g. using `history['ins'][img_idx].copy()`

schencej replied:
I believe the copy is not needed. The image is not mutated by the `visualize_saliency()` function. The `cv2.addWeighted()` function used there returns a new image.

brianhhu commented on examples/atari_deepRL_saliency.ipynb:
```diff
@@ -711,14 +688,11 @@
-    "img[35:195,:,0] += postprocess_sal(critic_sal_maps[frame_ix-1020], 600).astype('uint16')\n",
-    "img = img.clip(1,255).astype('uint8')\n",
+    "viz = visualize_saliency(img, actor_sal_maps[frame_ix-1020], critic_sal_maps[frame_ix-1020])\n",
     "\n",
-    "plt.imshow(img)\n",
+    "plt.imshow(viz)\n",
```
The saliency here looks much weaker compared to what we had originally, but it could be due to the fact that those visualization tricks were used to help "saturate" the saliency and make it stronger than it actually was. It could also be due to the fact that the frame we're explaining here is different than the original frame (see the comment above about the ATARI environment and seed), which I had cherry-picked to show good saliency.

schencej replied:
This was indeed caused by the incorrect gym version (see above).

Alpha blending the heatmap with the game frame results in slightly less visible saliency in the visualization. I increased the alpha value of the saliency slightly to make it pop a little more.

brianhhu reviewed (commented):
I added my comments in the review. Regarding naming of the new implementation, I would suggest something less RL-specific (citing the Greydanus, et al. '18 paper in the docstring still) since it is plausible that this scoring method could also be used for CV-based applications. Maybe something like `SquaredDifferenceScoring`?

Purg reviewed (commented):
I resolved a number of previous comments that look to be addressed. I don't have much else to add that hasn't already been commented on: jut a couple wording or nitpicky things that could be ignored for now or addressed in a different PR (so as to not block this one).

schencej commented on xaitk_saliency/impls/gen_classifier_conf_sal/policy_scoring.py:
```diff
@@ -0,0 +1,53 @@
+    Although the policy network is not an image classification model per se, it
+    shares certain similarities in that it also receives as input an image
+    (i.e. a video game frame) and outputs vectors of action and/or value scores
+    which are similar to classifier confidences.
```
I removed this blurb after renaming the implementation.

Purg commented on tests/impls/gen_classifier_conf_sal/test_squared_difference_scoring.py:
```diff
@@ -0,0 +1,91 @@
+import pytest
+
+
+class TestPolicyScoring:
```
This should be updated to reflect the class-rename.

schencej commented:
Should be good to go.

brianhhu reviewed (approved):
LGTM, thanks!

Purg reviewed (approved):
LGTM as well.

Purg commented:
Rebasing and updating this branch.
